### PR TITLE
DR-391 Add unit test action

### PR DESCRIPTION
**Pull request by timrickwood**
## Description
Prototype approach to unit tests and adding them to workflow

## Type of changes

- [ ] Refactoring (non-breaking change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [ ] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [ ] I have followed the code style of the project
- [ ] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [ ] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: .github/workflows/cicd-pipeline.yaml [hunk 1]
```diff
@@ -4,7 +4,7 @@ on:
   push:
   pull_request:
     branches:
-      - main
+      - task/DR-391_Unit_test_action
```
Comment (jope1) [inline .github/workflows/cicd-pipeline.yaml:7]: Are you going to refactor this stuff into our new develop workflow once my PR is merged?

**Review by jope1 — CHANGES_REQUESTED** [review]

File: application/.gitignore [hunk 2]
```diff
@@ -0,0 +1,7 @@
+coverage.xml
```
Comment (timrickwood) [inline application/.gitignore:1]: We have lots of gitignores and they look broadly the same ... can we get one that covers all sub-directories? Is that better ?

File: application/healthcare-services-data-manager/app.py [hunk 3]
```diff
@@ -1,83 +1,52 @@
 from chalice import Chalice
+import hs_service
```
Comment (timrickwood) [inline application/healthcare-services-data-manager/app.py:2]: Are we happy with the naming of this service ? Should we more explicit eg healthcareservices_service or hcs_service...

File: application/healthcare-services-data-manager/hs_service.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
+import boto3
+
+TABLE_NAME = "healthcare_services"
```
Comment (timrickwood) [inline application/healthcare-services-data-manager/hs_service.py:3]: Not for this ticket maybe ? Should we add some exception handling to this code

File: application/healthcare-services-data-manager/requirements.txt [hunk 5]
```diff
@@ -1,4 +1,6 @@
 Flask==1.1.4
-pytest==5.4.1
+# pytest==6.2.5
```
Comment (timrickwood) [inline application/healthcare-services-data-manager/requirements.txt:2]: Remove dead code

File: application/locations-data-manager/.gitignore [hunk 6]
```diff
@@ -0,0 +1,7 @@
+coverage.xml
```
Comment (timrickwood) [inline application/locations-data-manager/.gitignore:1]: See above comment about one gitignore to ignore them all

File: application/locations-data-manager/app.py [hunk 7]
```diff
@@ -1,83 +1,52 @@
 from chalice import Chalice
+import l_service
```
Comment (timrickwood) [inline application/locations-data-manager/app.py:2]: The name of this class/object - locations_service better ?

File: application/locations-data-manager/app.py [hunk 8]
```diff
@@ -1,83 +1,52 @@
...
-            "HospitalLocation": "London",
-        }
-    )
+    request = app.current_request.json_body
+    data = {
+        "id": request["id"],
+        "HospitalName": request["HospitalName"],
+        "HospitalLocation": request["HospitalLocation"],
+    }
+    print(data)
```
Comment (timrickwood) [inline application/locations-data-manager/app.py:27]: remove print

File: application/locations-data-manager/app.py [hunk 9]
```diff
@@ -1,83 +1,52 @@
...
-        }
-    )
+    print("Get locations record...")
+    request = app.current_request.json_body()
 
-    print(response["Item"])
-
-    print("\n\n\n-------\n\n\n")
+    l_id = request["id"]
+    print("Get l_id record...".l_id)
```
Comment (timrickwood) [inline application/locations-data-manager/app.py:14]: Remove print - replace with debug if needed

File: application/locations-data-manager/l_service.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
+import boto3
```
Comment (timrickwood) [inline application/locations-data-manager/l_service.py:1]: name of file ? locations_service ? 
Exception handling ?

File: application/organisation-affiliations-data-manager/.gitignore [hunk 11]
```diff
@@ -0,0 +1,7 @@
+coverage.xml
```
Comment (timrickwood) [inline application/organisation-affiliations-data-manager/.gitignore:1]: comment re gitignore

File: application/organisation-affiliations-data-manager/app.py [hunk 12]
```diff
@@ -1,83 +1,52 @@
...
-        }
-    )
+    print("Get organisationaffiliations record...")
+    request = app.current_request.json_body()
 
-    print(response["Item"])
-
-    print("\n\n\n-------\n\n\n")
+    oa_id = request["id"]
+    print("Get oa_id record...".oa_id)
```
Comment (timrickwood) [inline application/organisation-affiliations-data-manager/app.py:14]: Remove print replace with debug ?

File: application/organisation-affiliations-data-manager/app.py [hunk 13]
```diff
@@ -1,83 +1,52 @@
...
-            "HospitalLocation": "London",
-        }
-    )
+    request = app.current_request.json_body
+    data = {
+        "id": request["id"],
+        "HospitalName": request["HospitalName"],
+        "HospitalLocation": request["HospitalLocation"],
+    }
+    print(data)
```
Comment (timrickwood) [inline application/organisation-affiliations-data-manager/app.py:27]: remove print

File: application/organisation-affiliations-data-manager/app.py [hunk 14]
```diff
@@ -1,83 +1,52 @@
 from chalice import Chalice
+import oa_service
```
Comment (timrickwood) [inline application/organisation-affiliations-data-manager/app.py:2]: More explicit name of service ?

File: application/organisation-affiliations-data-manager/oa_service.py [hunk 15]
```diff
@@ -0,0 +1,44 @@
+import boto3
```
Comment (timrickwood) [inline application/organisation-affiliations-data-manager/oa_service.py:1]: Name of file/module - org_affiliations_service ?
Exception handling now or later

File: application/organisation-affiliations-data-manager/requirements.txt [hunk 16]
```diff
@@ -1,4 +1,6 @@
 Flask==1.1.4
-pytest==5.4.1
+# pytest==6.2.5
```
Comment (timrickwood) [inline application/organisation-affiliations-data-manager/requirements.txt:2]: Remove dead code

File: application/organisations-data-manager/.gitignore [hunk 17]
```diff
@@ -0,0 +1,7 @@
+coverage.xml
```
Comment (timrickwood) [inline application/organisations-data-manager/.gitignore:1]: one gitignore or many

File: application/organisations-data-manager/app.py [hunk 18]
```diff
@@ -1,83 +1,52 @@
 from chalice import Chalice
+import o_service
```
Comment (timrickwood) [inline application/organisations-data-manager/app.py:2]: organisations_service ?

File: application/organisations-data-manager/app.py [hunk 19]
```diff
@@ -1,83 +1,52 @@
...
-        }
-    )
+    print("Get organisations record...")
+    request = app.current_request.json_body()
 
-    print(response["Item"])
-
-    print("\n\n\n-------\n\n\n")
+    o_id = request["id"]
+    print("Get o_id record...".o_id)
```
Comment (timrickwood) [inline application/organisations-data-manager/app.py:14]: Replace print with debug if necessary ?

File: application/organisation-affiliations-data-manager/app.py [hunk 20]
```diff
@@ -1,83 +1,52 @@
...
-    print("Delete organisationaffiliations")
-
-    organisationaffiliations_table = dynamodb.Table("organisation_affiliations")
-
-    organisationaffiliations_table.delete_item(
-        Key={
-            "id": "002",
-        }
-    )
+    print("Delete healthcareservice record...")
```
Comment (timrickwood) [inline application/organisation-affiliations-data-manager/app.py:47]: If we keep this print correct it - not healthcare service in this instance

File: application/organisations-data-manager/app.py [hunk 21]
```diff
@@ -1,83 +1,52 @@
...
-            "HospitalLocation": "London",
-        }
-    )
+    request = app.current_request.json_body
+    data = {
+        "id": request["id"],
+        "HospitalName": request["HospitalName"],
+        "HospitalLocation": request["HospitalLocation"],
+    }
+    print(data)
```
Comment (timrickwood) [inline application/organisations-data-manager/app.py:27]: remover or replace with debug ?

File: application/organisations-data-manager/app.py [hunk 22]
```diff
@@ -1,83 +1,52 @@
...
-    print("Update organisations")
-
-    organisations_table = dynamodb.Table("organisations")
-
-    organisations_table.update_item(
-        Key={"id": "002"},
-        UpdateExpression="set HospitalLocation= :h",
-        ExpressionAttributeValues={":h": "York"},
-        ReturnValues="UPDATED_NEW",
+    print("Updating healthcareservice record...")
```
Comment (timrickwood) [inline application/organisations-data-manager/app.py:36]: If we keep this print correct it - not healthcare service in this instance

File: application/organisations-data-manager/app.py [hunk 23]
```diff
@@ -1,83 +1,52 @@
...
-    print("Delete organisations")
-
-    organisations_table = dynamodb.Table("organisations")
-
-    organisations_table.delete_item(
-        Key={
-            "id": "002",
-        }
-    )
+    print("Delete healthcareservice record...")
```
Comment (timrickwood) [inline application/organisations-data-manager/app.py:47]: If we keep this print correct it - not healthcare service in this instance

File: application/organisations-data-manager/o_service.py [hunk 24]
```diff
@@ -0,0 +1,44 @@
+import boto3
```
Comment (timrickwood) [inline application/organisations-data-manager/o_service.py:1]: Name of module ? organisations_service ?
Exception handling now or later

File: application/organisations-data-manager/requirements.txt [hunk 25]
```diff
@@ -1,4 +1,6 @@
 Flask==1.1.4
-pytest==5.4.1
+# pytest==6.2.5
```
Comment (timrickwood) [inline application/organisations-data-manager/requirements.txt:2]: Remvoe dead code

File: application/questionnaire-responses-data-manager/.gitignore [hunk 26]
```diff
@@ -0,0 +1,7 @@
+coverage.xml
```
Comment (timrickwood) [inline application/questionnaire-responses-data-manager/.gitignore:1]: one or many ?

File: application/questionnaire-responses-data-manager/app copy.py [hunk 27]
```diff
@@ -0,0 +1,52 @@
+from chalice import Chalice
+import qr_service
```
Comment (timrickwood) [inline application/questionnaire-responses-data-manager/app copy.py:2]: Name ?

File: application/questionnaire-responses-data-manager/requirements.txt [hunk 28]
```diff
@@ -1,4 +1,6 @@
 Flask==1.1.4
-pytest==5.4.1
+# pytest==6.2.5
```
Comment (timrickwood) [inline application/questionnaire-responses-data-manager/requirements.txt:2]: Remove dead code

File: application/questionnaires-data-manager/.gitignore [hunk 29]
```diff
@@ -0,0 +1,7 @@
+coverage.xml
```
Comment (timrickwood) [inline application/questionnaires-data-manager/.gitignore:1]: one or many

File: application/questionnaires-data-manager/app.py [hunk 30]
```diff
@@ -1,83 +1,52 @@
 from chalice import Chalice
+import q_service
```
Comment (timrickwood) [inline application/questionnaires-data-manager/app.py:2]: questionnaires_service?

File: application/questionnaires-data-manager/app.py [hunk 31]
```diff
@@ -1,83 +1,52 @@
...
-        }
-    )
+    print("Get questionnaires record...")
+    request = app.current_request.json_body()
 
-    print(response["Item"])
-
-    print("\n\n\n-------\n\n\n")
+    q_id = request["id"]
+    print("Get q_id record...".q_id)
```
Comment (timrickwood) [inline application/questionnaires-data-manager/app.py:14]: print or debug

File: application/questionnaires-data-manager/app.py [hunk 32]
```diff
@@ -1,83 +1,52 @@
...
-            "HospitalLocation": "London",
-        }
-    )
+    request = app.current_request.json_body
+    data = {
+        "id": request["id"],
+        "HospitalName": request["HospitalName"],
+        "HospitalLocation": request["HospitalLocation"],
+    }
+    print(data)
```
Comment (timrickwood) [inline application/questionnaires-data-manager/app.py:27]: print or debug

File: application/questionnaires-data-manager/q_service.py [hunk 33]
```diff
@@ -0,0 +1,44 @@
+import boto3
+
```
Comment (timrickwood) [inline application/questionnaires-data-manager/q_service.py:2]: name ?

File: application/questionnaires-data-manager/requirements.txt [hunk 34]
```diff
@@ -1,4 +1,6 @@
 Flask==1.1.4
-pytest==5.4.1
+# pytest==6.2.5
```
Comment (timrickwood) [inline application/questionnaires-data-manager/requirements.txt:2]: remove dead code

File: .github/workflows/pytest.yaml [hunk 35]
```diff
@@ -0,0 +1,23 @@
...
+          fetch-depth: 0
+
+      - name: Set up Python
+        uses: actions/setup-python@v4
+        with:
+          python-version: 3.9
+
+      - name: Run unit test script
+        run: |
+          export AWS_DEFAULT_REGION=us-west-2
```
[the remark this answers is not in the document]
Comment (jope1) [reply .github/workflows/pytest.yaml:22]: Agree; there is a github environment variable called AWS_REGION

File: application/healthcare-services-data-manager/hs_service.py (hunk 4 above)
Comment (jope1) [reply application/healthcare-services-data-manager/hs_service.py:3]: Might be an idea to raise a separate ticket to cover off exception handing, then someone else can have a go at that and writing the unit tests to cover it?

File: application/healthcare-services-data-manager/requirements.txt (hunk 5 above)
Comment (jope1) [reply application/healthcare-services-data-manager/requirements.txt:2]: How are we making sure that each service uses the same versions of imports?

File: application/locations-data-manager/app.py (hunk 9 above)
Comment (jope1) [reply application/locations-data-manager/app.py:14]: Have we thought about logging levels yet and a way to configure them?

**Review by jope1 — CHANGES_REQUESTED** [review]

**Review by timrickwood — COMMENTED** [review]
I can't approve or request changes cos I raised the original PR

**Review by gdcNHS — APPROVED** [review]

**Review by jope1 — APPROVED** [review]

**Pull request merged**